## 1. Introduction

Different parts of the genome have their own evolutionary history and geneology. However, while we are evaluating the demography and selective pressure of a population, we usually make our inference based on the overall statistics of the genome. The heterogeniety of evolution in the genome is usually ignored. Thus, we need to investigate the evolutionary property in different regions, such as coding region, introns or untranslated regions. We started from the site requency spectrum and the summary statistics including nucleotide diversity (pi) and Tajima's D.

## 2. Material and Methods

We utalized the Illumina reads from 15 strains of DSPR and analyzed the SNPs in **vcf** files. Our workflow is composed of five sections, read mapping, SNP calling plus annotation, making coordinates, classifying SNPs and statistics analysis. Each section was done with one bash script. The detailed explanation of the scripts were documented in the `Readme.md` file. Here, we provided an overview of our concepts. Each strain was sequenced three time and 45 filtered **fastq** files were mapped against the reference genome independently. Since the read length were 54 bp long, we had to use `bwa aln` and `bwa sampe`. Three **bam** files of each strain were merged together. Afterwards, 15 **bam** were used for SNP calling and then annotated independently. For further analysis, these 15 **vcf** files were merged into one **vcf** file. (marked as "over all-1" later)

We aimed to categorize the locations of SNPs according to their coordinates and to exclude overlapping multiply annotated regions. We artificially classified the locations of SNPS into 8 categories, including:

1. long introns (li; longer than 120 bp)
2. Short introns (si; shorter than 65bp)
3. distal intergenic regions (dig; at least 4000 bp away)
4. proximal intergenic regions (pig)
5. five prime untranslated regions (five_prime_UTR)
6. three prime untranslated regions (three_prime_UTR)
7. synonymous (syn)
8. nonsynonymous (nonsyn)

The coordinats of genetic regions were extracted from **gff** file deposited at Flybase. Because five_prime_UTR, three_prime_UTR, introns and exons overlapped with each other sometimes, the coordinate of one of them were intersected with the rest. Intergenic regions were defined as regions outside genes (`bedtools complement "gene.bed"`) and splited into distal and proximal parts. Introns were classfiled as short or long introns and 66-bp to 120-bp long introns were discarded. Then, we used the **bed** files to extract SNPs from the merged **vcf** file. The resulted **vcf** files were used for downstream analysis.

Finally, theta, pi and allele frquencies were calculated by `vcftools`. The Tajima's D was calculated by vcf-kit. The site frequency spectrum was made with `bioawk` and `R`. To visualize our results, `ggplot2` in `R` was used. Only values on chromosome 2L, 2R, 3L,3R, 4 and X were calculated.

To optimize the analysis, computational threads were used as many as allowed. Some of the steps were constructed with pipelines to save some time. Overall, the SNP calling was the most time-consuming. One large **bam** file was difficult for ordinary single-threaded `Freebayes` to call the SNPs sccessfully. We used the parallized version of `Freebayes` and annotated it with `snpEff` again. (marked as "over all-2" later)

## Results

All the pipelines worked well. A **bam** file from one read was around 2 GB and merged **bam** file was around 6 GB. A raw **vcf** file generated by `Freebayes` was around 520 MB while a **vcf.gz** file had only 20% of the original size. After filtering, the **vcf.gz** file (of each srain) retained only 55% of the size. Finally the annotated **vcf** files were around 620 MB. The **bed** files were small (less than 1 MB), as well as **fai**, **bai** and **tbi** files. The summary of pi and Tajima's D in each type of SNPs were listed in the Table 1, 2 and 3. 


#### **Table.1 Nucleotide diversity (pi) from different categories of SNPs.**  The window size was 10000 bp. "over all-Tony" were SNPs from all categories and the data were provided by Anthony Long. "over all-1" refereed to the SNPs from all categories in a merged vcf file. "over all-2" referred to to the SNPs from all categories in a one-time-called vcf file. "A1" was a homogenuos strain from DSPR.

|category| mean| standard deviation |
|---|---|---|
|over all-Tony|3.680e-03|2.03e-03|
|over all-1|1.161e-04|1.04e-04|
|over all-2|4.358e-03|2.42e-03|
|A1|1.848e-04|1.818e-04|
|syn|2.344e-05|3.275e-05|
|nonsyn|9.827e-06| 2.178e-05|
|li|8.735e-05|8.667e-05|
|si|7.153e-06|9.573e-06|
|5' UTR|9.377e-06|1.241e-05|
|3' UTR|1.937e-05|2.664e-05|
|dig|8.081e-05|8.411e-05|
|pig|5.543e-05|6.327e-05|

#### **Table.2 Tajima's D from different categories of SNPs.** The window size was 10000 bp. "over all-Tony" were SNPs from all categories and the data were provided by Anthony Long. "over all-1" refereed to the SNPs from all categories in a merged vcf file. "over all-2" referred to to the SNPs from all categories in a one-time-called vcf file. "A1" was a homogenuos strain from DSPR.

|category| mean| standard deviation |
|---|---|---|
|over all-Tony|0.259|0.68|
|over all-1|1.069|1.05|
|over all-2|0.124|0.679|
|A1|not applicable|not applicable|
|syn|0.401|1.052|
|nonsyn|0.854|1.056|
|li|0.828|1.008|
|si|0.243|1.021|
|5' UTR|0.559|1.099|
|3' UTR|0.653|1.045|
|dig|0.853|1.014|
|pig|0.704|1.043|

#### Figure.1 **Some examples of dotplots**
|pi|Tajima's D|
|---|---|
|<img src="r6.annotated.pi .png" width="400" height="400">|<img src="r6.tajimasd.vk .png" width="400" height="400">|
|<img src="synSNPs.pi .png" width="400" height="400">|<img src="synSNPs.tajimasd.vk .png" width="400" height="400">|

#### Figure.2 **Some examples of the site frequency spectrum**
|||
|:---|:---|
|<img src="nonsynSNPs.freq2 .png" width="400" height="400">|<img src="r6.SNPs.freq2 .png" width="400" height="400">|
|||
|<img src="synSNPs.freq2 .png" width="400" height="400">|<img src="merged.annotated.freq2 .png" width="400" height="400">|
|||

## Discussion

This time, our pipeine worked but the statistics turned incorrect. For example, the Tajima's D were usualy smaller than 0 (doi:10.1038/nature04107). We tried to identify the problem, and initially suspected the manner of merging **vcf** file was not ideal. Thus, other independently called SNPs were tested here. If we called the SNPs all at once ("over all-2"), the results turned closer to the real world data but there were still some difference. The results derived from "over all-2" were similar to the results derived from the **vcf** file provided by Anthody Long. In addition, the nucleotide diversity (pi) of A1 strain was also extremely low when compared to the over-all results from 15 strains. We think the SNP caller worked fine so we need to review the filtering and quality control while generating the data. If this does not work, we might need to investigate the mapping results.